syntax = "proto3";

package relationalai.lqp.v1;

import "relationalai/lqp/v1/fragments.proto";
import "relationalai/lqp/v1/logic.proto";

message Transaction {
  repeated Epoch epochs = 1;
  Configure configure = 2;
}

message Configure {
  int64 semantics_version = 1;
}

message Epoch {
  repeated Write persistent_writes = 1;
  repeated Write local_writes = 2;
  repeated Read reads = 3;
}

//
// Write operations
//

message Write {
  oneof write_type {
    Define define = 1;
    Undefine undefine = 2;
    Context context = 3;
  }
}

message Define {
  Fragment fragment = 1;
}

message Undefine {
  FragmentId fragment_id = 1;
}

message Context {
  repeated RelationId relations = 1;
}

//
// Read operations
//

message Read {
  oneof read_type {
    Demand demand = 1;
    Output output = 2;
    WhatIf what_if = 3;
    Abort abort = 4;
  }
}

message Demand {
  RelationId relation_id = 1;
}

message Output {
  string name = 1;
  RelationId relation_id = 2;
}

message WhatIf {
  string branch = 1;
  Epoch epoch = 2;
}

message Abort {
  string name = 1;
  RelationId relation_id = 2;
}
